scene/2d/tile_map_layer.cpp

```diff
@@ -1691,8 +1691,8 @@ void TileMapLayer::_notification(int p_what) {
 
 		case NOTIFICATION_EXIT_TREE: {
 			dirty.flags[DIRTY_FLAGS_LAYER_IN_TREE] = true;
-			// Update immediately on exiting, and force cleanup.
-			_internal_update(true);
+			// Update immediately on exiting, but do not force cleanup.
```

Comment:
Why this change? It's necessary, please explain the change as it affects *far* more than just the scenes, this is still not correct IMO

Reply:
I took a look at your repro with the light occluder, and it looks like while force cleaning the atlas tiles did solve it, scene tiles never had that issue, and should not have been changed to force clean when the TileMap is removed from the tree. 

Force cleaning the scene tile nodes when the TileMap is removed from the tree erases any state those nodes have accumulated since being instantiated and causes them to be reloaded in their default state when the TileMap is eventually re-added to the tree.

For a solution that solves both issues, I wish I could just take the force_cleanup out of _scenes_update, but that causes other problems. Now that the scene tile nodes haven't been wiped out, the rest of _scenes_update tries to access them in a bad non-deferred way (which is part of why I originally wanted to switch to use _queue_internal_update).

To do this right, I think perhaps I will either have to split up the atlas/scene tile update code paths a bit more, or split up the force_clean/update code paths a bit more.

Reply:
It did so before so it wasn't changed by this, it was restored to how it was done before the thread fixes, my PR restored the behavior before that that had been done for a long time, so this change shouldn't be made like this, please restore it and we can work with a solution down the line, but this change is invalid

Again, scene tiles *were* deleted when exiting the tree for a long time, but that behavior was accidentally removed for a short while after thread fixes were applied, the reason for the forced cleanup is that `is_inside_tree` returns *`false`* when exiting the tree, so it worked because the deferred call. But the deferred call was thread unsafe

The problem is that there'd be no way to solve this without failing to remove the scene tiles even when cleaning up at destruction, since this was the behavior previously I'd say it's invalid to change it this way, unless it's safe to not force cleaning up when deleting the node itself

But it should be done without touching the forced cleanup of other parts

Reply:
The fact that the updates break things if the force is delayed implies this is a more involved problem and that the clearing of the tiles is assumed and built into the process IMO, I'm not convinced it's a bug, but a possible limitation that might need more workarounds and other solutions, and at least might not be appropriate for the 4.3 release given that it risks causing bugs and issues

For the forced part a simple solution would be to add two arguments, or to restore the old variable tracking if we're in the destructor and free them only then, but I'm not sure it's trivial to implement this change without a lot of other changes, all making this fragile

But I think we should wait for the opinion of groud on this

But I feel that using scene tiles should be adjusted to adapt to this, by having them handle states in a way that handles this, rather than keeping them around, it makes a lot of processing with the tiles more manageable I think

We could also implement some way to track and retain state, but I don't know that scene tiles being stateful is something that's guaranteed to be supported, we don't allow providing per-tile options to them for example, there's a proposal for that I believe

Reply:
Definitely, I recognize the change is invalid as-is, and I'm not proposing that we submit it yet. I super don't want to re-introduce the issue that you linked.

It sounds like you're saying that you think maybe scene tiles nodes _should_ be deleted and re-instantiated when the TileMap exits/re-enters the tree. If that is true, then you're right that this is not a bug, and we can abandon trying to fix this PR.

However, that seems at odds with [Godot's tree model](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-remove-child:~:text=The%20node%2C%20along%20with%20its%20children%2C%20are%20not%20deleted.) to delete and re-instantiate those nodes when their parent is simply added/removed. Happy to wait for others' opinions on this as well.

If we aren't going to fix this issue, I think this behavior should be documented [here](https://docs.godotengine.org/en/stable/classes/class_tilesetscenescollectionsource.html#class-tilesetscenescollectionsource:~:text=Scenes%20are%20instantiated%20as%20children%20of%20the%20TileMap%20when%20it%20enters%20the%20tree.), and that the child nodes that are created should maybe be changed to be ["Internal"](https://docs.godotengine.org/en/stable/classes/class_node.html#enum-node-internalmode) children of the TileMap.

Reply:
> However, that seems at odds with [Godot's tree model](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-remove-child:~:text=The%20node%2C%20along%20with%20its%20children%2C%20are%20not%20deleted.) to delete and re-instantiate those nodes when their parent is simply added/removed. Happy to wait for others' opinions on this as well.

That's arguably different, these aren't normal node children but internally generated nodes, handled by the node itself.

But let's see what the maintainers for this part says on the desired behavior, or the limitations we should follow. The solution should handle things safely so must be adjusted to the change here, so depends on the assumptions that was made for the rest of the code, which seems to be that the scenes should be removed on exiting the tree, and the question is if that's desired or not, but the behavior is not fully accidental, it might be a case of copy-pasting some of the code, but the condition for deletion is explicit.